src/modules/cmdpal/Microsoft.CmdPal.UI.ViewModels/ContextMenuViewModel.cs

```diff
@@ -0,0 +1,267 @@
+﻿// Copyright (c) Microsoft Corporation
+// The Microsoft Corporation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.ObjectModel;
+using CommunityToolkit.Mvvm.ComponentModel;
+using CommunityToolkit.Mvvm.Messaging;
+using Microsoft.CmdPal.UI.ViewModels.Messages;
+using Microsoft.CommandPalette.Extensions;
+using Microsoft.CommandPalette.Extensions.Toolkit;
+using Microsoft.Diagnostics.Utilities;
+using Windows.System;
+
+namespace Microsoft.CmdPal.UI.ViewModels;
+
+public partial class ContextMenuViewModel : ObservableObject,
+    IRecipient<UpdateCommandBarMessage>,
+    IRecipient<OpenContextMenuMessage>
+{
+    public ICommandBarContext? SelectedItem
+    {
+        get => field;
+        set
+        {
+            if (field != null)
+            {
+                field.PropertyChanged -= SelectedItemPropertyChanged;
+            }
+
+            field = value;
+            SetSelectedItem(value);
+
+            OnPropertyChanged(nameof(SelectedItem));
+        }
+    }
+
+    [ObservableProperty]
+    private partial ObservableCollection<List<IContextItemViewModel>> ContextMenuStack { get; set; } = [];
+
+    private List<IContextItemViewModel>? CurrentContextMenu => ContextMenuStack.LastOrDefault();
+
+    [ObservableProperty]
+    public partial ObservableCollection<IContextItemViewModel> FilteredItems { get; set; } = [];
+
+    [ObservableProperty]
+    public partial bool FilterOnTop { get; set; } = false;
+
+    private string _lastSearchText = string.Empty;
+
+    public ContextMenuViewModel()
+    {
+        WeakReferenceMessenger.Default.Register<UpdateCommandBarMessage>(this);
+        WeakReferenceMessenger.Default.Register<OpenContextMenuMessage>(this);
+    }
+
+    public void Receive(UpdateCommandBarMessage message)
+    {
+        SelectedItem = message.ViewModel;
+    }
+
+    public void Receive(OpenContextMenuMessage message)
+    {
+        FilterOnTop = message.ContextMenuFilterLocation == ContextMenuFilterLocation.Top;
+
+        ResetContextMenu();
+
+        OnPropertyChanging(nameof(FilterOnTop));
```

Comment:
Does setting `FilterOnTop` not also raise these events for us? I thought that was the point of `[ObservableProperty]`